### PR TITLE
Fix/search http client

### PR DESCRIPTION
**Pull request by sully90**
### What

Switch to CloseableHttpClient (and responses)

### How to review

Code review

### Who can review

Anyone but me

### CODE REVIEW

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import java.net.URISyntaxException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class for building Closable Http clients to be used by Babbage
+ */
+public class BabbageHttpClient implements AutoCloseable {
+
+    protected final CloseableHttpClient httpClient;
+    protected final URI HOST;
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Theses should be lower case and should be separate i.e. 
```
protected final uri;
protected final host;
```

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    private final PoolingHttpClientConnectionManager connectionManager;
+    private final IdleConnectionMonitorThread monitorThread;
+
+    public BabbageHttpClient(String host, ClientConfiguration configuration) {
+        HOST = resolveHostUri(host);
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Be consistent use `this` on the left hand of all assignments to the object in the constructor.

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        this.connectionManager = new PoolingHttpClientConnectionManager();
+        HttpClientBuilder customClientBuilder = HttpClients.custom();
+        configure(customClientBuilder, configuration);
+        httpClient = customClientBuilder.setConnectionManager(connectionManager)
+                .build();
+        this.monitorThread = new IdleConnectionMonitorThread(connectionManager);
+        this.monitorThread.start();
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Add a log to say starting monitor thread.

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            builder.setPort(givenHost.getPort());
+            builder.setPath(givenHost.getPath());
+            builder.setUserInfo(givenHost.getUserInfo());
+        } else {
+            builder.setScheme("http");
+            builder.setHost(host);
+        }
+        try {
+            return builder.build();
+        } catch (URISyntaxException e) {
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Add log message

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        System.out.println("Shutting down connection pool to host:" + HOST);
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Replace with logger.

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        httpClient.close();
+        System.out.println("Successfully shut down connection pool");
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: ditto

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        private HttpClientConnectionManager connMgr;
+
+        public IdleConnectionMonitorThread(HttpClientConnectionManager connMgr) {
+            super();
+            this.connMgr = connMgr;
+        }
+
+        @Override
+        public void run() {
+            System.out.println("Running connection pool monitor");
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: ditto

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            try {
+                while (!shutdown) {
+                    synchronized (this) {
+                        wait(5000);
+                        // Close expired connections every 5 seconds
+                        connMgr.closeExpiredConnections();
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: maybe add a log message here

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                        // Close connections
+                        // that have been idle longer than 30 sec
+                        connMgr.closeIdleConnections(60, TimeUnit.SECONDS);
+                    }
+                }
+            } catch (InterruptedException ex) {
+                System.err.println("Connection pool monitor failed");
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: replace with logger

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                ex.printStackTrace();
+            }
+        }
+
+        public void shutdown() {
+            System.out.println("Shutting down connection pool monitor");
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: ditto

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+    private class ShutdownHook extends Thread {
+        @Override
+        public void run() {
+            try {
+                if (httpClient != null) {
+                    close();
+                }
+            } catch (Exception e) {
+                System.err.println("Falied shutting down http client for, " + HOST);
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: ditto

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+                notifyAll();
+            }
+        }
+
+    }
+
+    private class ShutdownHook extends Thread {
+        @Override
+        public void run() {
+            try {
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: Add log message here

File: src/main/java/com/github/onsdigital/babbage/util/http/PooledHttpClient.java [hunk 13]
```diff
@@ -257,65 +150,9 @@ private String getErrorMessage(HttpEntity entity) {
             String s = EntityUtils.toString(entity);
             return s;
         } catch (Exception e) {
-            logEvent(e).error("error attempting to read HTTPEntity as string");
+            System.err.println("Failed reading content service:");
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/PooledHttpClient.java]: use logger

File: src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    protected final URI host;
+    private final PoolingHttpClientConnectionManager connectionManager;
+    private final IdleConnectionMonitorThread monitorThread;
+
+    public BabbageHttpClient(String host, ClientConfiguration configuration) {
+        this.host = resolveHostUri(host);
+        this.connectionManager = new PoolingHttpClientConnectionManager();
+        HttpClientBuilder customClientBuilder = HttpClients.custom();
+        configure(customClientBuilder, configuration);
+        httpClient = customClientBuilder.setConnectionManager(connectionManager)
```
Comment (daiLlew) [inline src/main/java/com/github/onsdigital/babbage/util/http/BabbageHttpClient.java]: add `this` to httpclient assignment.

File: src/main/java/com/github/onsdigital/babbage/search/external/SearchEndpoints.java [hunk 15]
```diff
@@ -1,11 +1,11 @@
 package com.github.onsdigital.babbage.search.external;
 
-import com.github.onsdigital.babbage.search.external.requests.search.requests.ListType;
+import com.github.onsdigital.babbage.search.external.requests.search.ListType;
 
 public enum SearchEndpoints {
 
     SEARCH("/search/"),
-    SEARCH_ONS("/search/%s/"),
+    SEARCH_ONS("/search/"),
```
Comment (janderson2) [inline src/main/java/com/github/onsdigital/babbage/search/external/SearchEndpoints.java]: Do we still need `SEARCH_ONS`, can we not just use `SEARCH` now that the proxying is gone?

File: src/main/java/com/github/onsdigital/babbage/configuration/ExternalSearch.java [hunk 16]
```diff
@@ -16,6 +16,7 @@
     private static final String EXTERNAL_SEARCH_PORT_KEY = "EXTERNAL_SEARCH_PORT";
     private static final String EXTERNAL_SEARCH_ENABLED_KEY = "ENABLE_SEARCH_SERVICE";
     private static final String SEARCH_NUM_EXECUTORS_KEY = "SEARCH_NUM_EXECUTORS";
+    private static final String SEARCH_MAX_CONNECTIONS = "SEARCH_MAX_CONNECTIONS";
```
[the remark this answers is not in the document]
Comment (sully90) [reply src/main/java/com/github/onsdigital/babbage/configuration/ExternalSearch.java]: See https://github.com/ONSdigital/dp-ci/pull/541

File: src/main/java/com/github/onsdigital/babbage/search/external/SearchEndpoints.java (hunk 15 above)
Comment (sully90) [reply src/main/java/com/github/onsdigital/babbage/search/external/SearchEndpoints.java]: Removed SEARCH_ONS entry

**Pull request merged**